File: backend/internal/decoder/type_detector.go
package decoder

import (
	"regexp"
	"strings"
)

type QRResultType string

const (
	QRResultTypeURL     QRResultType = "url"
	QRResultTypeEmail   QRResultType = "email"
	QRResultTypePhone   QRResultType = "phone"
	QRResultTypeWifi    QRResultType = "wifi"
	QRResultTypeText    QRResultType = "text"
	QRResultTypeUnknown QRResultType = "unknown"
)

var (
	emailPattern = regexp.MustCompile(`^[^\s@]+@[^\s@]+\.[^\s@]+$`)
	phonePattern = regexp.MustCompile(`^\+?[0-9][0-9\s().-]{6,}$`)
	digitPattern = regexp.MustCompile(`[0-9]`)
)

func DetectType(value string) QRResultType {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return QRResultTypeUnknown
	}

	lower := strings.ToLower(trimmed)

	switch {
	case strings.HasPrefix(lower, "http://") || strings.HasPrefix(lower, "https://"):
		return QRResultTypeURL
	case strings.HasPrefix(lower, "mailto:") || emailPattern.MatchString(trimmed):
		return QRResultTypeEmail
	case strings.HasPrefix(lower, "tel:") || isPhone(trimmed):
		return QRResultTypePhone
	case strings.HasPrefix(trimmed, "WIFI:"):
		return QRResultTypeWifi
	default:
		return QRResultTypeText
	}
}

func isPhone(value string) bool {
	if !phonePattern.MatchString(value) {
		return false
	}

	return len(digitPattern.FindAllString(value, -1)) >= 7
}

File: backend/src/qr/type-detector.ts
import { QRResultType } from "./dto/extract-response.dto";

const emailPattern = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
const phonePattern = /^\+?[0-9][0-9\s().-]{6,}$/;
const digitPattern = /[0-9]/g;

export function detectQRValueType(value: string): QRResultType {
  const trimmed = value.trim();
  if (!trimmed) {
    return "unknown";
  }

  const lower = trimmed.toLowerCase();

  if (lower.startsWith("http://") || lower.startsWith("https://")) {
    return "url";
  }

  if (lower.startsWith("mailto:") || emailPattern.test(trimmed)) {
    return "email";
  }

  if (lower.startsWith("tel:") || isPhone(trimmed)) {
    return "phone";
  }

  if (trimmed.startsWith("WIFI:")) {
    return "wifi";
  }

  return "text";
}

function isPhone(value: string): boolean {
  if (!phonePattern.test(value)) {
    return false;
  }

  return (value.match(digitPattern) ?? []).length >= 7;
}
